lib/Conversion/SeqToSV/FirRegLowering.h

```diff
@@ -10,26 +10,127 @@
 #ifndef CONVERSION_SEQTOSV_FIRREGLOWERING_H
 #define CONVERSION_SEQTOSV_FIRREGLOWERING_H
 
+#include "circt/Dialect/Comb/CombOps.h"
 #include "circt/Dialect/HW/HWOps.h"
 #include "circt/Dialect/SV/SVOps.h"
 #include "circt/Dialect/Seq/SeqOps.h"
 #include "circt/Support/LLVM.h"
 #include "circt/Support/Namespace.h"
 #include "circt/Support/SymCache.h"
+#include "mlir/IR/Visitors.h"
+#include <mlir/IR/ValueRange.h>
+#include <stack>
+#include <unordered_set>
 
 namespace circt {
+
+using namespace hw;
+// This class computes the set of muxes that are reachable from an op.
+// The heuristic propagates the reachability only through the 3 ops, mux,
+// array_create and array_get. All other ops block the reachability.
+// This analysis is built lazily on every query.
+// The query: is a mux is reachable from a reg, results in a DFS traversal
+// of the IR rooted at the register. This traversal is completed and the result
+// is cached in a Map, for faster retrieval on any future query of any op in
+// this subgraph.
+class ReachableMuxes {
+public:
+  ReachableMuxes(HWModuleOp m) : module(m) {}
+
+  bool isMuxReachableFrom(Operation *regOp, Operation *muxOp) {
+    return llvm::any_of(regOp->getResult(0).getUsers(), [&](Operation *user) {
+      if (opBlocksReachability(user))
+        return false;
+      buildReachabilityFrom(user);
+      return reachableMuxes[user].contains(muxOp);
+    });
+  }
+
+private:
+  static inline bool opBlocksReachability(Operation *op) {
+    return (!isa<comb::MuxOp, ArrayGetOp, ArrayCreateOp>(op));
+  }
+  void buildReachabilityFrom(Operation *startNode) {
+    // This is a backward dataflow analysis.
+    // First build a graph rooted at the `startNode`. Every user of an operation
+    // that doesnot block the reachability is a child node. Then, the ops that
+    // are reachable from a node is computed as the union of the Reachability of
+    // all its child nodes.
+    // for all child in the Children(node)
+    // Reachability(node) = node + Union{Reachability(child)}
+    if (visited.find(startNode) != visited.end())
+      return;
+    // The op and its users information that needs to be tracked on the stack
+    // for an iterative DFS traversal.
+    struct OpUserInfo {
+      Operation *op;
+      const mlir::ResultRange::user_range userRange;
+      mlir::ResultRange::user_iterator userIter;
+
+      OpUserInfo(Operation *op)
+          : op(op), userRange(op->getUsers()), userIter(userRange.begin()) {}
+
+      // Increments the itertor to the next valid user op and returns false if
+      // the iterator reaches the end of the range.
+      auto getNextValid(mlir::ResultRange::user_iterator &iter) const {
+        for (; iter != userRange.end(); ++iter)
+          if (!opBlocksReachability(*iter))
+            return true;
+        return false;
+      }
+    };
+    // The stack to record enough information for an iterative post-order
+    // traversal.
+    std::stack<OpUserInfo> stk;
```

Comment:
Just curious, what's the motivation for stdlib data structures for the stack and visited ops when we often use LLVM data structures?

Reply:
Was an oversight, switched to llvm data structures.